lib/Dialect/ESI/cosim/cosim_dpi_server/DpiEntryPoints.cpp

```diff
@@ -0,0 +1,207 @@
+//===- DpiEntryPoints.cpp - ESI cosim DPI calls -----------------*- C++ -*-===//
+//
+// Cosim DPI function implementations. Mostly C-C++ gaskets to the C++
+// RpcServer.
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Dialect/ESI/cosim/Server.h"
+#include "circt/Dialect/ESI/cosim/dpi.h"
+
+#include <algorithm>
+#include <cstdlib>
+#include <iostream>
+
+using namespace circt::esi::cosim;
+
+static RpcServer *server = nullptr;
+
+// ---- Helper functions ----
+
+/// Get the TCP port on which to listen. Defaults to 0xECD (ESI Cosim DPI), 3789
+/// in decimal.
+int getPort() {
+  const char *portEnv = getenv("COSIM_PORT");
+  if (portEnv == nullptr) {
+    printf("[COSIM] RPC server port not found. Defaulting to 3789.\n");
+    return 0xECD;
+  }
+  printf("[COSIM] Opening RPC server on port %s\n", portEnv);
+  return std::strtoull(portEnv, nullptr, 10);
+}
+
+/// Check that an array is an array of bytes and has some size.
+// NOLINTNEXTLINE(misc-misplaced-const)
+static int validateSvOpenArray(const svOpenArrayHandle data,
+                               int expectedElemSize) {
+  if (svDimensions(data) != 1) {
+    printf("DPI-C: ERROR passed array argument that doesn't have expected 1D "
+           "dimensions\n");
+    return -1;
+  }
+  if (svGetArrayPtr(data) == NULL) {
+    printf("DPI-C: ERROR passed array argument that doesn't have C layout "
+           "(ptr==NULL)\n");
+    return -2;
+  }
+  int totalBytes = svSizeOfArray(data);
```

Comment:
should these use size_t?

Reply:
Nope. svSizeOfArray() returns an int. (It's one of those standard function calls in the DPI spec.)